src/metabase/driver/sql_jdbc/connection.clj

```diff
@@ -166,7 +190,7 @@
   "Return an appropriate JDBC connection spec to test whether a set of connection details is valid (i.e., implementing
   `can-connect?`)."
   [driver details]
-  (let [details-with-tunnel (ssh/include-ssh-tunnel details)]
+  (let [details-with-tunnel (driver/incorporate-ssh-tunnel-details driver details)]
```

Comment:
this `incorporate-ssh-tunnel-details` being a multimethod seems strange to me. As it works right now, all sql based stuff gets it, but not mongo and others. The logic doesn't seem like it would change on the type of driver: if there's a tunnel if it's open we're good, otherwise re-open it. that seems fundamental and not to depend on the type of database.

Reply:
also, its name gives me no indication that it will actually reopen an ssh tunnel.

Reply:
The logic might change per driver - some drivers are RESTful APIs instead of JDBC.

Reply:
how would that change a metabase managed ssh connection though?

Reply:
Refer to the H2 implementation.  That's why I made it a multimethod, so that it could be supported in H2.  Most of the JDBC DBs use `:host` and `:port` separately, but H2 actually just has `:db` which contains the full connection string.  When the tunnel is opened, the port portion of that URL has to be swapped out dynamically for the tunnel entry port.

Reply:
I see. I read that on this a while back and assumed it was just validation of tcp rather than seeing it hot-swapping the port. I think maybe the version you have for `:sql-jdbc` should be moved to the driver namespace as `:default` so everything gets that except for h2. Don't see any reason its just sql-jdbc stuff.

Reply:
I think my rationale was that the `driver.clj` file seemed to be fairly sparse, and everything that used the tunnel previously had `:sql-jdbc` as a parent.  Moving the default implementation there would end up importing the `ssh` namespace and a bunch of stuff there.  But if that makes more sense, I'm happy to make the change.

Reply:
what else would be imported? For sure ssh but it makes sense for the driver stuff to depend on ssh. But i think it makes the most sense for all connections to get this behavior unless they have a particular way to do it (like h2)

Reply:
Hmm, actually it seems this introduces a cyclic dependency.

`Cyclic load dependency: [ /metabase/driver ]->/metabase/util/ssh->[ /metabase/driver ]->/metabase/plugins/lazy_loaded_driver->/metabase/plugins/initialize->/metabase/plugins`

Might be a mess to untangle. 😢

Reply:
i don't think so. this change added the metabase.driver dep into ssh. so just move it and restart the repl and you should be good to go.

Reply:
Good point.  Backed that out, but different error of the same kind.

```
Syntax error compiling at (metabase/driver/util.clj:1:1).
Cyclic load dependency: [ /metabase/driver ]->/metabase/driver/util->/metabase/public_settings->/metabase/util/ssh->[ /metabase/driver ]->/metabase/plugins/lazy_loaded_driver->/metabase/plugins/initialize->/metabase/plugins
```

Reply:
i think that's still a remnant of your broken state. it's saying that m.u.ssh depends on driver which is no longer true right?

Reply:
I don't think so, now.  Looking on `master`...

- `ssh` depends on `public-settings`
- `public-settings` depends on `driver.util`
- `driver.util` depends on `driver`

Therefore, we can't add a dependency on `ssh` to `driver` or else we end up with this cycle.

Reply:
ahh. i didn't have that change that added the dependency on public-settings. i'm sorry about that. I guess to leave this as `:sql-jdbc`. Looking a bit further, its only called from the jdbc stuff anyways so we can do it this way for now. I apologize for not reading closer @jeff303 !